src/frames.rs

```diff
@@ -0,0 +1,86 @@
+use std::fmt::{self, Formatter, Display};
+use std::num::ParseIntError;
+use std::ops::{Add, AddAssign, Div, Sub};
+use std::str::FromStr;
+use std::time::Duration;
+
+use crate::sample_rate::SampleRate;
+
+#[derive(Clone, Copy, PartialOrd, PartialEq, Ord, Eq)]
+pub struct Frames(u64);
+
+fn scale(value: u64, max: u64) -> u32 {
+    assert!(max < (std::u32::MAX as u64));
+    (value * (std::u32::MAX as u64) / max) as u32
+}
+
+impl Frames {
+    pub const fn new(value: u64) -> Frames {
+        Frames(value)
+    }
+
+    pub const fn as_usize(&self, frame_size: usize) -> usize {
+        (self.0 as usize) * frame_size
+    }
+
+    pub const fn from_usize(size: usize, frame_size: usize) -> Frames {
+        Frames((size / frame_size) as u64)
+    }
+}
+
+impl From<Frames> for u64 {
+    fn from(frames: Frames) -> u64 {
+        frames.0
+    }
+}
+
+impl From<u64> for Frames {
+    fn from(value: u64) -> Frames {
+        Frames(value)
+    }
+}
+
+impl Add for Frames {
```

Comment:
You can have a look at https://crates.io/crates/shrinkwraprs as that would give you Add, AddAssign and Sub for free.

Reply:
Yeah, I actually looked at that package after our IRL discussion, but as far as I can tell (I might be missing something) it doesn't actually implement any of those traits, but rather gives you `Deref`.

This means, as far as I understand, that I can't actually e.g. `a + b` and get a `Frames`, result. Rather it will deref `a` & `b` to `u64`s, which means that the result will be a `u64` instead of `Frames`.

I didn't look too much into it though, so maybe I was just missing something.

Reply:
Yeah you would need to wrap the result once again to retain the Frames type. Not ideal but easy derive for type structs is not here yet sadly :(

Reply:
I wonder how hard it would be to just derive all the `std::ops::*` traits directly? It shouldn't be too hard right?

You should be able to just forward directly to the underlying types implementation, e.g. for `Add`:

```rust
impl Add for $struct {
    type Output = $struct;

    fn add(self, other: $struct) -> $struct {
        $struct($primitive::add(self.0, other.0))
    }
}
```

Where `$struct` would be `Frames` and `$primitive` `u64` in this case